### DISCUSSION TITLE
Flaky test due to missing assets

### DISCUSSION THREAD

**Discussion by robbevp** [ElMassimo/vite_ruby#357] [Q&A] [answered]
Hi there 👋 

I've been seeing some of our tests flake, due to vite ruby not being able to find the assets.
This seems to happen every so often either locally or in CI. When re-running the job, the problem disappears.

The file is obviously there, as this code is running locally and in production without issues.

Any ideas to what I might be missing here? I've added an example error below (with some details anonymised), but this seems to appear on different controller tests without much pattern to it.

```
Error:
Admin::SomeControllerTest#some_tests:
ActionView::Template::Error: Vite Ruby can't find images/logo.png in public/vite-test/manifest.json or public/vite-test/manifest-assets.json.

Possible causes:
  - The file path is incorrect.
  - The file is not in the entrypoints directory.
  - Some files are outside the sourceCodeDir, and have not been added to watchAdditionalPaths.

Visit the Troubleshooting guide for more information:
  https://vite-ruby.netlify.app/guide/troubleshooting.html#troubleshooting


Content in your manifests:
{
}

Last build in test mode:
{
  "success": true,
  "errors": "",
  "timestamp": "2023-04-20 14:10:15",
  "vite_ruby": "3.3.0",
  "digest": "9e52797a355f5a4d1319b63a8e24733786d64f25"
}


    app/components/admin/shared/sidebar_component.html.erb:46:in `call'
    app/views/layouts/admin.html.erb:24
    app/views/layouts/admin.html.erb:16
 ```

**Comment by ElMassimo** [comment]
Hi Robbe!

I'd suggest trying to debug why the `manifest.json` file is missing or empty, as the error shows:

```
Content in your manifests:
{
}
```

**Comment by gurgeous** [comment, marked as answer]
I can chime in here - in our case this was caused by [running tests in parallel](https://edgeguides.rubyonrails.org/testing.html#parallel-testing). We've been having intermittent errors as well and I finally dug into it this weekend... It happens when multiple tests trigger the vite build at the same time. Of course, the next time you run tests the vite build will be up to date and things work perfectly.

I can make it happen by running this:

```
$ rm -rf public/vite-test && rake test
```

We are now trying to mitigate in several ways:

- turning off parallelization in CI
- manually running `vite build` before tests (we use [just](https://github.com/casey/just) so this is easy)
- I might try this in our Procfile too - https://vite-ruby.netlify.app/guide/development.html#hmr-for-integration-tests-%E2%9C%85. This has the added benefit of speeding things up when devs run tests.

@ElMassimo any other clever ideas? Might be worth a mention in the docs at least. We love Vite Ruby!

**Reply by ElMassimo** [reply]
Hi Adam! Thanks for sharing a potential cause for this error.

Related discussion:

- https://github.com/ElMassimo/vite_ruby/discussions/136#discussioncomment-4466983

Extended the _Development_ guide to include an [_Integration Tests in the CI_ section](https://vite-ruby.netlify.app/guide/development.html#integration-tests-in-the-ci-%E2%9C%85), right after [_HMR for Integration Tests_](https://vite-ruby.netlify.app/guide/development.html#hmr-for-integration-tests-%E2%9C%85).

**Reply by robbevp** [reply]
Hi there,

Parallel testing seems to be the root of the problem indeed. Thanks for your help!
